### DISCUSSION TITLE
Environment-based setups

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#186] [BC Patterns]
In case of integration with other system, usually we work with "endpoints" and "API Calls".  However, in test-environments, it's important to NOT call the LIVE endpoint and vica versa.

In a world where we expect to be able to copy environments (hence, also copy endpoints to test-environments), these settings can turn into undeliberate catastrophies.

Therefore, this pattern does:
- have logic to bind a certain URL (GetUrl) to a certain Environment-code
- setup-page that binds a certain setup to a certain environment
- Setup.Get() shouldn't work anymore, because of the multiple setups - but is replace by Setup.Get(GetCurrEnvironmentCode());

This way, an accidental use of a live endpoint is avoided .. and an automatic switch to the test-endpoint is facilitated.

**Comment by JeremyVyska** [comment]
Ooo, clever!

**Reply by waldo1001** [reply]
In fact - the platform-part is already a library-app on AppSource .. not sure that anyone would be able to use that  (since it's only part of our products as library apps 🤔) ... I should test that..

**Reply by kine** [reply]
I think it cannot be used until we will install your solution... you will need to publish it as separate AppSource app and make it free...

**Reply by waldo1001** [reply]
Hm .. I could do that 🤔

**Comment by TheDoubleH** [comment]
As someone who have seen emails sent to customer when someone played in a test system etc. this idea isn't just for api calls - it's for a lot of things.

We already have IsSandbox and IsProduction in tenant management. (See 
https://demiliani.com/2018/07/04/dynamics-365-business-central-how-to-detect-if-your-extension-is-running-on-a-sandbox/ ) or are you thinking even more separation? (QA sandbox vs DEV Sandbox) 

But I like where this is going.

**Comment by pibcht** [comment]
@waldo1001, I just tested successfully something that can improve this part :
> Setup.Get() shouldn't work anymore, because of the multiple setups - but is replace by Setup.Get(GetCurrEnvironmentCode());

You can totally use polymorphism by redeclaring Get function.
On this example, I use it to get randomly one line

```al
// on table
keys
{
    key(Key1; PK) { Clustered = true; }
}

[TryFunction]
procedure Get()
begin
    Case Random(3) of
        1: Get('A');
        2: Get('B');
        3: Get('C');
    end;
end;
```

```al
// elsewhere
MyTable.Get();
Message('%1\%2', MyTable.PK, MyTable.Data);
```

[MyTable.Table.txt](https://github.com/microsoft/alguidelines/files/8774002/MyTable.Table.txt)

**Reply by waldo1001** [reply]
Hm - interesting that it works.  However, I'm not sure overloading the Get is recommended 😬.  What or others' opinions?

May be foresee a method like `Rec.GetForCurrEnvironment()` ? 🤔

**Reply by kine** [reply]
I will rather not do it becaus it could easilly lead to never-ending loop if you by accident use Get() inside the overloaded Get() (when it grow complex with events etc.)

**Comment by JeremyVyska** [comment]
Oh!  I meant to add this here.   There's also something we should be doing in events & procedures for External Connectors, which is checking to make sure the Execution Context isn't Install/Upgrade.  See this thread:

https://twitter.com/JeremyVyska/status/1575015526992326656
